File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null){
            return ans;
        }

        Deque<TreeNode> q = new ArrayDeque<>();
        q.add(root);

        while(!q.isEmpty()){
            List<Integer> temp = new ArrayList<>();

            int sz = q.size();
            for(int i = sz; i > 0; i--)
            {
                TreeNode currNode = q.poll();

                temp.add(currNode.val);

                if(currNode.left != null){
                    q.add(currNode.left);
                }
                if(currNode.right != null){
                    q.add(currNode.right);
                }
            }
            ans.addFirst(temp);
        }
        return ans;
    }
}

File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, vector<vector<int>>& ans, int lvl)
    {
        if(root == NULL){
            return ;
        }

        if(ans.size() <= lvl){
            ans.push_back({});
        }

        ans[lvl].push_back(root->val);
        lvl++;

        dfs(root->left, ans,lvl);
        dfs(root->right,ans,lvl);

    }
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        ios::sync_with_stdio(false);
        cin.tie(NULL);

        if(root==NULL){
            return {};
        }
        vector<vector<int>> ans;
        dfs(root, ans,0);

        reverse(ans.begin(), ans.end());

        return ans;
    }
    // vector<vector<int>> levelOrderBottom(TreeNode* root) {
        
    // }
};